### PR TITLE
UP-4711: smartLdapGroupStore improvements

### PR DESCRIPTION
**Pull request by jgribonvald**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [ ] tests are included
-   [ ] documentation is changed or added
-   [ ] [message properties][] have been updated with new phrases
-   [ ] view conforms with [WCAG 2.0 AA][]

##### Description of change

This permit to set a new king of ldap groups loading.
Docs + test will follow in some month but I can provide a description in conversation.
This code works for us and somes other since several years in production.

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

**Comment by jgribonvald** [comment]
Some docs for the use case:

**Description of group structure**
```
ou=people,dc=org,dc=fr
     uid=XXXXX,ou=peoples,dc=org,dc=fr 
           uid=XXXXX
           cn,...
           isMemberOf=leaf1:subleaf1:group1
           isMemberOf=leaf1:subleaf1:group2
           ismemberOf=.....
           isMemberOf=leaf1:subleaf5:groupx

ou=groups,,dc=org,dc=fr
     cn=leaf1:subleaf1:group1,ou=groups,,dc=org,dc=fr
           cn=leaf1:subleaf1:group1
           member=uid=XXXXX,ou=peoples,dc=org,dc=fr
           member= cn=leaf1:subleaf5:groupx,ou=groups,,dc=org,dc=fr
           member=....(dn only of peoples or groups)
           hasMember=leaf1:subleaf5:groupx
           hasMember= XXXXX
           hasMember=...(id key, so uid for people object or cn for group object)      
           
```
like you can see each groups is a flat tree representation, and we recompose the tree into uPortal by decomposition with the ":" char, it could be any other char (but must be forbidden in groups names)

** Example of configuration for a such use case **
```xml
 <bean id="smartLdapGroupStore" class="org.apereo.portal.groups.smartldap.SmartLdapGroupStore" lazy-init="false" depends-on="defaultLdapContext">

        <property name="ldapContext" ref="defaultLdapContext"/>

        <property name="memberOfAttributeName" value="isMemberOf"/>

        <property name="baseGroupDn" value="ou=groups"/>

        <property name="childGroupKeyRegex" value="cn=(.*),ou=groups,${ldap.defaultLdapContext.baseDn}"/>

        <property name="filter" value="(objectClass=groupOfNames)"/>

        <property name="groupsTreeRefreshIntervalSeconds" value="${smartldapgroupstore.groupsTreeRefreshIntervalSeconds:900}"/>

        <property name="resolveMemberGroups" value="false"/>
        <property name="resolveDn" value=""/><!--Used with resolveMemberGroups -->

        <property name="attributesMapper">
            <bean class="org.apereo.portal.groups.smartldap.SimpleAttributesMapper">
                <property name="keyAttributeName">
                    <value>cn</value>
                </property>
                <property name="groupNameAttributeName">
                    <value>cn</value>
                </property>
                <property name="membershipAttributeName">
                    <value>member</value>
                </property>
            </bean>
        </property>
    </bean>
```
We can also resolve peoples members into groups, but it consume a lot of resources on big number of groups and users (400k users into 45k groups, so it's disable for us)

**Comment by jgribonvald** [comment]
These change should not break anything and could work at least with openldap memberOf overlay.

For test case I would like to provide an embedded LDAP with unboundid-ldap-sdk https://ldap.com/unboundid-ldap-sdk-for-java/ like I already done with a custom portlet to retrieve organizational informations and cache it (when you manage more than 250 organizations), 
here is an example of implementation: https://github.com/GIP-RECIA/esco-ChangeEtablissement/commit/92e76122d3f32395a48094c05714ea9ee619cb28

The only thing is that it doesn't simulate an Active Directory, but seemingly tools could be found. And these change should be tested on before merging (I can't make test on).

Any help in testing on native use and against Active Directory would be appreciated.

**Comment by jgribonvald** [comment]
For previous réference see https://github.com/Jasig/uPortal/pull/698

File: uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java [hunk 1]
```diff
@@ -630,6 +690,61 @@ private GroupsTree buildGroupsTree() {
...
+        // Process each group in order to create the groups that will represent the Grouper's
+        // folders
+        // and maintain the consistency regarding the membership.
+        for (String groupKey : newGroups.keySet()) {
+
+            boolean stop = false;
+            boolean isAFolder = false;
+            String currentPath = groupKey;
+            while (!stop) {
+                final String containingFolder = getContainingFolder(currentPath);
```
Comment (ChristianMurphy) [inline uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: If I'm reading this correctly, current path will be a `:` separated list of paths, and this loops until is has processed all folders.
Would something like `for (String containingFolder : groupKey.split(":")) {` work?

Comment (jgribonvald) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: This should work. This code come from a Java 1.5 version and I migrate only needed parts...

Comment (jgribonvald) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: Finally this won't works as the id of the group is the entire path and the containing folder is the entire path, so we need to remove only the last part each time we want the containing folder.

As example if you have this group:
Organizations:org_a:students:class_A
the first containing folder is:
Organizations:org_a:students
the next containing folder is:
Organizations:org_a
and the last containing folder is:
Organizations

So you need to make it in this way, with your solution you will need to rebuild the id path on each values.

Comment (ChristianMurphy) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: I understand. :thinking:
My main concern is with `while (!stop) {`, using a flag as a `while` conditional tends to be a code smell. :nose: 
What would be the implications of something like:

```java
var groupKeys = groupKey.split(":");
while (groupKeys.size() > 0) {
  var containingFolder = String.join(":", groupKeys);

  // logic goes here

  groupKeys.removeLast();
}
```

Comment (jgribonvald) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: I'm agree the entire code of the loop could be rewrited ;)  I Will watch it when I will be back from holidays (I can't before I try to solve this issue https://github.com/uPortal-contrib/uPortal-web-components/issues/43)

Comment (jgribonvald) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: what do you think about that:
```java

for (String groupKey : newGroups.keySet()) {
            String currentPath = groupKey;

            // The group that represents the grouper folder is created.
            final IEntityGroup folderGroup =
                new EntityTestingGroupImpl(currentPath, IPerson.class);
            folderGroup.setCreatorID("System");
            folderGroup.setName(currentPath);
            folderGroup.setDescription("Grouper Folder");
            newFolders.put(currentPath, folderGroup);

            String containingFolder = this.getContainingFolder(currentPath);
            while (!containingFolder.isEmpty()) {
                // Adds the containing folder as a parent of this current path
                List<String> parentsList = newParents.get(currentPath);
               ...
                childrenList.add(currentPath);

                // The remaining prefix of the group has already been processed.
                if (newFolders.containsKey(containingFolder)) {
                   break; 
                }
                currentPath = containingFolder;

                containingFolder = this.getContainingFolder(currentPath);
            }
        }
```

Comment (ChristianMurphy) [reply uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java]: That could work :+1:

**Comment by jgribonvald** [comment]
Requested change done and tested !

File: uPortal-webapp/src/main/resources/org/apereo/portal/groups/smartldap/init.crn [hunk 2]
```diff
@@ -42,28 +43,75 @@
...
+		            record = new LdapRecord(record.getGroup(), newChildren);
+	            }
+	            GROUPS.add(record);
+
+	            //   Parses the group names in order to handle the Grouper folders.
+
+                String current = record.getGroup().getName();
+                int sepPos;
+                boolean stop = false;
+	            boolean isGroup = true;
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/resources/org/apereo/portal/groups/smartldap/init.crn]: Indented different than line 88?

Comment (jgribonvald) [reply uPortal-webapp/src/main/resources/org/apereo/portal/groups/smartldap/init.crn]: the indent is not well managed with crn files ...

**Comment by jgribonvald** [comment]
review on file indent !

**Pull request merged**